dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/utils/TaskCacheUtils.java

```diff
@@ -0,0 +1,160 @@
+     * @return cache key
+     */
+    public static String generateCacheKey(TaskInstance taskInstance, TaskExecutionContext taskExecutionContext) {
+        List<String> keyElements = new ArrayList<>();
+        keyElements.add(String.valueOf(taskInstance.getTaskCode()));
+        keyElements.add(String.valueOf(taskInstance.getTaskDefinitionVersion()));
+        keyElements.add(String.valueOf(taskInstance.getIsCache().getCode()));
+        keyElements.add(getTaskInputVarPoolData(taskInstance, taskExecutionContext));
+        String data = StringUtils.join(keyElements, "_");
+        return DigestUtils.sha256Hex(data);
+    }
+
+    /**
+     * generate cache key for task instance which is cache execute
+     * this key will record which cache task instance will be copied, and cache key will be used
+     * tagCacheKey = sourceTaskId + "-" + cacheKey
+     * @param sourceTaskId source task id
+     * @param cacheKey cache key
+     * @return tagCacheKey
+     */
+    public static String generateTagCacheKey(Integer sourceTaskId, String cacheKey) {
+        return sourceTaskId + MERGE_TAG + cacheKey;
+    }
+
+    /**
+     * revert cache key tag to source task id and cache key
+     * @param tagCacheKey cache key
+     * @return Pair<Integer, String>, first is source task id, second is cache key
+     */
+    public static Pair<Integer, String> revertCacheKey(String tagCacheKey) {
+        Pair<Integer, String> taskIdAndCacheKey;
+        if (tagCacheKey == null) {
+            taskIdAndCacheKey = Pair.of(-1, "");
+            return taskIdAndCacheKey;
+        }
+        if (tagCacheKey.contains(MERGE_TAG)) {
+            String[] split = tagCacheKey.split(MERGE_TAG);
+            if (split.length == 2) {
+                taskIdAndCacheKey = Pair.of(Integer.parseInt(split[0]), split[1]);
+            } else {
+                taskIdAndCacheKey = Pair.of(-1, "");
+            }
+            return taskIdAndCacheKey;
+        } else {
+            return Pair.of(-1, tagCacheKey);
+        }
+    }
+
+    /**
+     * get hash data of task input var pool
+     * there are two parts of task input var pool: from upstream task and workflow global parameters
+     * @param taskInstance task instance
+     * taskExecutionContext taskExecutionContext
+     */
+    public static String getTaskInputVarPoolData(TaskInstance taskInstance, TaskExecutionContext context) {
+        JsonNode taskParams = JSONUtils.parseObject(taskInstance.getTaskParams());
+
+        // The set of input values considered from localParams in the taskParams
+        Set<String> propertyInSet = JSONUtils.toList(taskParams.get("localParams").toString(), Property.class).stream()
+                .filter(property -> property.getDirect().equals(Direct.IN))
+                .map(Property::getProp).collect(Collectors.toSet());
+
+        // The set of input values considered from `${var}` form task definition
+        propertyInSet.addAll(getScriptVarInSet(taskInstance));
```

Comment:
Should we also consider environment variables, such as adding environment config as part of cache key element?

Reply:
Any task definition-related changes(including environment code, but it doesn't include the dolphinscheduler_env.sh) have a task version to capture the changes

Reply:
> Should we also consider environment variables, such as adding environment config as part of the cache key element?

Is the `environment variables` mean  dolphinscheduler_env.sh?

Reply:
I've noticed that you generate cache key with `taskDefinitionVersion`, `taskDefinition` only has field `environment_code`, however, environment config change will not change `environment_code`.

Reply:
Dose `environment config` means the actual definition content about `environment_code `?

Reply:
> Dose environment config means the actual definition content about environment_code ?

Yes, it is. I think it's better consider it when generate cache key in this PR for two reasons:
- `EnvironmentConfig` can be directly fetched from either `TaskInstance` or `TaskExecutionContext`.
- `EnvironmentConfig` is like `taskCode` which DS takes charge of their management. Rather, data from the worker or from the resource center may change, user themselves should take into consideration.

WDYT?

Reply:
Yes, I have added `EnvironmentConfig` field to generate the cache key.